core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java

```diff
@@ -53,68 +58,99 @@
             throw new ElasticsearchException("Field [" + suggestionContext.getField() + "] is not a completion suggest field");
         }
-        final IndexReader indexReader = searcher.getIndexReader();
         CompletionSuggestion completionSuggestion = new CompletionSuggestion(name, suggestionContext.getSize());
         spare.copyUTF8Bytes(suggestionContext.getText());
-
         CompletionSuggestion.Entry completionSuggestEntry = new CompletionSuggestion.Entry(new StringText(spare.toString()), 0, spare.length());
         completionSuggestion.addTerm(completionSuggestEntry);
-
-        String fieldName = suggestionContext.getField();
-        Map<String, CompletionSuggestion.Entry.Option> results = Maps.newHashMapWithExpectedSize(indexReader.leaves().size() * suggestionContext.getSize());
-        for (LeafReaderContext atomicReaderContext : indexReader.leaves()) {
-            LeafReader atomicReader = atomicReaderContext.reader();
-            Terms terms = atomicReader.fields().terms(fieldName);
-            if (terms instanceof Completion090PostingsFormat.CompletionTerms) {
-                final Completion090PostingsFormat.CompletionTerms lookupTerms = (Completion090PostingsFormat.CompletionTerms) terms;
-                final Lookup lookup = lookupTerms.getLookup(suggestionContext.fieldType(), suggestionContext);
-                if (lookup == null) {
-                    // we don't have a lookup for this segment.. this might be possible if a merge dropped all
-                    // docs from the segment that had a value in this segment.
-                    continue;
-                }
-                List<Lookup.LookupResult> lookupResults = lookup.lookup(spare.get(), false, suggestionContext.getSize());
-                for (Lookup.LookupResult res : lookupResults) {
-
-                    final String key = res.key.toString();
-                    final float score = res.value;
-                    final Option value = results.get(key);
-                    if (value == null) {
-                        final Option option = new CompletionSuggestion.Entry.Option(new StringText(key), score, res.payload == null ? null
-                                : new BytesArray(res.payload));
-                        results.put(key, option);
-                    } else if (value.getScore() < score) {
-                        value.setScore(score);
-                        value.setPayload(res.payload == null ? null : new BytesArray(res.payload));
-                    }
+        Map<Integer, Option> results = new LinkedHashMap<>(suggestionContext.getSize());
+        TopSuggestDocsCollector collector = new TopSuggestDocsCollector(suggestionContext.getSize());
+        suggest(searcher, toQuery(suggestionContext), collector);
+        for (TopSuggestDocs.SuggestScoreDoc suggestDoc : collector.get().scoreLookupDocs()) {
+            // TODO: currently we can get multiple entries with the same docID
```

Comment:
No not yet, but I am working on this one. Dedup'ing entries with early termination in the `TopSuggestDocsCollector` seems tricky. One way is to terminate as soon as we collect `num` unique doc ids from the `TopNSearcher` but still collect `num * maxEntriesPerDoc` entries in the priority queue. On `TopSuggestDocsCollector.get` merge the entries with the same doc id and present only `num` `SuggestScoreDocs`. Any thoughts on this?